src/fides/api/alembic/migrations/versions/1e750936a141_migrate_to_fideslang_2_0_uses_and_.py

```diff
@@ -0,0 +1,289 @@
+            )
+            bind.execute(
+                update_data_use_query,
+                {"policy_id": row["id"], "updated_rules": json.dumps(rules)},
+            )
+
+
+#####################
+## Data Categories ##
+#####################
+# The `key` is the old value, the `value` is the new value
+data_category_upgrades: Dict[str, str] = {
+    "user.credentials": "user.authorization.credentials",  # Verified in 2.0
+    "user.observed": "user.behavior",  # Verified in 2.0
+    "user.browsing_history": "user.behavior.browsing_history",  # Verified in 2.0
+    "user.media_consumption": "user.behavior.media_consumption",  # Verified in 2.0
+    "user.search_history": "user.behavior.search_history",  # Verified in 2.0
+    "user.organization": "user.contact.organization",  # Verified in 2.0
+    "user.non_specific_age": "user.demographic.age_range",  # Verified in 2.0
+    "user.date_of_birth": "user.demographic.date_of_birth",  # Verified in 2.0
+    "user.gender": "user.demographic.gender",  # Verified in 2.0
+    "user.political_opinion": "user.demographic.political_opinion",  # Verified in 2.0
+    "user.profiling": "user.demographic.profile",  # Verified in 2.0
+    "user.race": "user.demographic.race_ethnicity",  # Verified in 2.0
+    "user.religious_belief": "user.demographic.religious_belief",  # Verified in 2.0
+    "user.sexual_orientation": "user.demographic.sexual_orientation",  # Verified in 2.0
+    "user.financial.account_number": "user.financial.bank_account",  # Verified in 2.0
+    "user.genetic": "user.health_and_medical.genetic",  # Verified in 2.0
+}
+data_category_downgrades: Dict[str, str] = {
+    value: key for key, value in data_use_upgrades.items()
+}
+
+
+def update_datasets_data_categories(
+    bind: Connection, data_label_map: Dict[str, str]
+) -> None:
+    """Upgrade the datasets in the database to use the new data categories."""
+
+    # Get all datasets out of the database
+    existing_datasets: ResultProxy = bind.execute(
+        text("SELECT id, data_categories, collections FROM ctl_datasets;")
+    )
+
+    for row in existing_datasets:
+        # Update data categories at the top level
+        labels: Optional[List[str]] = row["data_categories"]
+
+        if labels:
+            updated_labels: List[str] = [
+                data_label_map.get(label, label) for label in labels
+            ]
+
+            update_label_query: TextClause = text(
+                "UPDATE ctl_datasets SET data_categories = :updated_labels WHERE id= :dataset_id"
+            )
+            bind.execute(
+                update_label_query,
+                {"dataset_id": row["id"], "updated_labels": updated_labels},
+            )
+
+        # Update the collections objects
+        collections: str = json.dumps(row["collections"])
+
+        for key, value in data_label_map.items():
+            collections = collections.replace(key, value)
```

Comment:
This string replace can cause unexpected behavior.  For example, say an old data category is `user.browsing_history`. I upgrade, and that becomes `user.behavior.browsing_history`.  I downgrade and that becomes the incorrect `user.observed.browsing_history`, because `user.behavior` was replaced with `user.observed`.  Suggestions are changing the order of the data categories dictionary - more specific first and more broad second, or to convert to a dictionary and do more targeted upgrades instead of string replace

Reply:
I'm not really worried about the downgrade case...would it be better to remove the downgrades altogether then to prevent it from causing bugs? Great catch here btw, I didn't think about the downgrade case

Reply:
Can we just move `"user.observed": "user.behavior",  # Verified in 2.0` to the last entry in the dictionary - looking closer, that's probably the only one? I think we need to get better at supporting downgrades generally

Reply:
Actually, can we back up and just make sure that this upgrade shouldn't be more targeted?

For customers that have built off of the taxonomy, say customer is using custom data category: `user.search_history.apples`.  Because we're doing string replacement, not exact matches in our migration,  this becomes, `user.behavior.search_history.apples`.  It is not clear to me that this is expected from a customer perspective.  @adamsachs

Reply:
With regards to blunt string replacement...

This is desired behavior according to @SteveDMurphy , as we don't want someone with an extended taxonomy to suddenly have orphaned elements

Reply:
A couple follow-up questions then, with an understanding that folks have ended up with orphaned elements in the past: 1) why are there orphaned elements if we don't touch the taxonomy itself? I was under the impression that the seed script added to DataCategory and DataUse and not updated. 2). On the other hand, if "blunt string replacement" is in fact desirable, would we need more updates to this migration? Broader string replacement in more locations than just collections on datasets.

Reply:
1) The default taxonomy is completely truncated and re-seeded every time. This causes the orphaning
2) That's a great point, let me look here and convert to `replace` instead of dict lookups where it makes sense

Reply:
@pattisdr I went through and updated the logic for find/replaces

Reply:
i'll be taking a closer look at all of this locally with some manual testing, but just wanted to ask about this in case you see it before i get to that:

> The default taxonomy is completely truncated and re-seeded every time

from a quick look at `seed.py`, i don't see where that truncation is done? i'll try this out locally and debug further to confirm, but all i see are upserts of the new taxonomy elements, which presumably do not _remove_ any of the outdated taxonomy elements?

Reply:
good catch Adam, it does seem like old taxonomy elements are not actually being removed.  So we need to get on the same page regarding whether they _should_ be removed here and if so, make that happen.

Reply:
I remembered that we were dropping taxonomy elements but it seems I was mistaken/we no longer are...

If we don't drop the existing taxonomy elements, we'll need to update them to all show `is_default=False`. This should fix the `version` validation issue.

Reply:
OK we reached some general consensus in [slack](https://ethyca.slack.com/archives/CBKK8TS74/p1693323605019939), it seems like we do want to:
- remove the deprecated taxonomy elements (probably should be done here within the migration itself?)
- update _all_ references to those taxonomy elements in-place (i.e. the blunt string replace)
- give some steps for a "fallback" plan upon upgrade if desired, i.e. don't automigrate, take a db backup and migrate manually